File: neetcode150/medium/22-Generate-Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> list:
        res = []
        def gen(o, c, temp):
            if o == n and c == n:
                res.append(temp)
                return
            if o < n:
                gen(o + 1, c, temp + "(")
            if c < o:
                gen(o, c + 1, temp + ")")
        
        gen(0, 0, "")
        return res

File: neetcode150/medium/cpp/22-Generate-Parentheses.cpp
#include <string>

using namespace std;

class Solution {
public:
    vector<string> all;
    void genParen(string curr, int left, int right) {
        if (right == 0) {
            all.push_back(curr);
            return;
        }
        if (left > 0) {
            genParen(curr + "(", left - 1, right);
        }
        if (left < right) {
            genParen(curr + ")", left, right - 1);
        }
    }
    vector<string> generateParenthesis(int n) {
        genParen("", n, n);
        return all;
    }
};